src/unwinder/arch/mips.rs

```diff
@@ -0,0 +1,397 @@
+            Self::F20 => Some("$f20"),
+            Self::F21 => Some("$f21"),
+            Self::F22 => Some("$f22"),
+            Self::F23 => Some("$f23"),
+            Self::F24 => Some("$f24"),
+            Self::F25 => Some("$f25"),
+            Self::F26 => Some("$f26"),
+            Self::F27 => Some("$f27"),
+            Self::F28 => Some("$f28"),
+            Self::F29 => Some("$f29"),
+            Self::F30 => Some("$f30"),
+            Self::F31 => Some("$f31"),
+            _ => return None,
+        }
+    }
+}
+
+// Match DWARF_FRAME_REGISTERS in libgcc
+pub const MAX_REG_RULES: usize = 188;
+
+#[repr(C)]
+#[derive(Clone, Default)]
+pub struct Context {
+    pub gp: [usize; 32],
+    #[cfg(target_feature = "single-float")]
+    pub fp: [usize; 32],
+}
+
+impl fmt::Debug for Context {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut fmt = fmt.debug_struct("Context");
+        for i in 0..=31 {
+            fmt.field(MIPS::register_name(Register(i as _)).unwrap(), &self.gp[i]);
+        }
+        #[cfg(target_feature = "single-float")]
+        for i in 32..=63 {
+            fmt.field(
+                MIPS::register_name(Register(i as _)).unwrap(),
+                &self.fp[i - 32],
+            );
+        }
+        fmt.finish()
+    }
+}
+
+impl ops::Index<Register> for Context {
+    type Output = usize;
+
+    fn index(&self, reg: Register) -> &usize {
+        match reg {
+            Register(0..=31) => &self.gp[reg.0 as usize],
+            #[cfg(target_feature = "single-float")]
+            Register(32..=63) => &self.fp[(reg.0 - 32) as usize],
+            _ => unimplemented!(),
+        }
+    }
+}
+
+impl ops::IndexMut<gimli::Register> for Context {
+    fn index_mut(&mut self, reg: Register) -> &mut usize {
+        match reg {
+            Register(0..=31) => &mut self.gp[reg.0 as usize],
+            #[cfg(target_feature = "single-float")]
+            Register(32..=63) => &mut self.fp[(reg.0 - 32) as usize],
+            _ => unimplemented!(),
+        }
+    }
+}
+
+macro_rules! code {
+    (save_gp) => {
+        "
+        sw $zero,   0x00($sp)
+        sw $ra,     0x7C($sp)
+        sw $gp,     0x70($sp)
+        sw $v0,     0x08($sp)
+        sw $v1,     0x0C($sp)
+        sw $a0,     0x10($sp)
+        sw $a1,     0x14($sp)
+        sw $a2,     0x18($sp)
+        sw $a3,     0x1C($sp)
+        sw $t0,     0x20($sp)
+        sw $t1,     0x24($sp)
+        sw $t2,     0x28($sp)
+        sw $t3,     0x2C($sp)
+        sw $t4,     0x30($sp)
+        sw $t5,     0x34($sp)
+        sw $t6,     0x38($sp)
+        sw $t7,     0x3C($sp)
+        sw $s0,     0x40($sp)
+        sw $s1,     0x44($sp)
+        sw $s2,     0x48($sp)
+        sw $s3,     0x4C($sp)
+        sw $s4,     0x50($sp)
+        sw $s5,     0x54($sp)
+        sw $s6,     0x58($sp)
+        sw $s7,     0x5C($sp)
+        sw $t8,     0x60($sp)
+        sw $t9,     0x64($sp)
+        "
```

Comment:
Where's the saving code for sp?

Reply:
Darn, I meant to move it and made a blunder, will fix

Reply:
This is not saving the correct sp. You have already clobbered it. It should be t0.